File: java/base/src/main/vsdk/toolkit/environment/geometry/geometricProcessing/polyhedralBoundedSolidOperators/_PolyhedralBoundedSolidSetOperatorSectorClassificationOnVertex.java
//= References:                                                             =
//= [MANT1988] Mantyla Martti. "An Introduction To Solid Modeling",         =
//=     Computer Science Press, 1988.                                       =

package vsdk.toolkit.environment.geometry.geometricProcessing.polyhedralBoundedSolidOperators;

import vsdk.toolkit.common.linealAlgebra.Vector3Dd;
import vsdk.toolkit.environment.geometry.volume.polyhedralBoundedSolid.nodes._PolyhedralBoundedSolidHalfEdge;

/**
This class is used to store vertex / halfedge neigborhood information for the
vertex/vertex classifier as proposed on section [MANT1988].15.5. and program
[MANT1988].15.6.
*/
public class _PolyhedralBoundedSolidSetOperatorSectorClassificationOnVertex
    extends _PolyhedralBoundedSolidOperator
    implements Comparable<_PolyhedralBoundedSolidSetOperatorSectorClassificationOnVertex>
{
    public _PolyhedralBoundedSolidHalfEdge he;
    public Vector3Dd ref1;
    public Vector3Dd ref2;
    public Vector3Dd ref12;
    public Vector3Dd referenceLine;
    public Vector3Dd referenceU;
    public Vector3Dd referenceV;
    public boolean wide;

    public _PolyhedralBoundedSolidSetOperatorSectorClassificationOnVertex()
    {
        referenceLine = null;
        referenceU = null;
        referenceV = null;
    }

    public double getAngle()
    {
        if ( referenceLine == null || referenceU == null || referenceV == null ) {
            return -1000;
        }

        double x;
        double y;
        double an;
        Vector3Dd a = ref1;

        if ( _PolyhedralBoundedSolidSetOperator.colinearVectorsWithDirection(ref1, referenceLine) ) {
            a = ref2;
        }

        Vector3Dd u;
        Vector3Dd v;

        u = new Vector3Dd(referenceU);
        u = u.normalized();
        v = new Vector3Dd(referenceV);
        v = v.normalized();
        a = a.normalized();

        x = a.dotProduct(u);
        y = a.dotProduct(v);

        an = Math.acos(x);
        if ( y < 0 ) an *= -1;

        return an;
    }

    public String toString()
    {
        String msg;

        msg = "R1: " + ref1 + " R2: " + ref2 + " HE " + he.startingVertex.id + "/" + he.next().startingVertex.id + (wide?"(W)":"(nw)");

        return msg;
    }

    public int compareTo(_PolyhedralBoundedSolidSetOperatorSectorClassificationOnVertex other)
    {
        double a;
        double b;

        a = this.getAngle();
        b = other.getAngle();

        if ( a > b) return 1;
        if ( a < b) return -1;
        return 0;
    }
}

File: typescript/base/src/main/vsdk/toolkit/environment/geometry/geometricProcessing/polyhedralBoundedSolidOperators/booleans/classification/_PolyhedralBoundedSolidSetOperatorSectorClassificationOnVertex.ts
//= References:                                                             =
//= [MANT1988] Mantyla Martti. "An Introduction To Solid Modeling",         =
//=     Computer Science Press, 1988.                                       =

import { Vector3Dd } from "../../../../../../common/linealAlgebra/Vector3Dd.js";
import type { _PolyhedralBoundedSolidHalfEdge } from "../../../../volume/polyhedralBoundedSolid/nodes/_PolyhedralBoundedSolidHalfEdge.js";
import { _PolyhedralBoundedSolidOperator } from "../../_PolyhedralBoundedSolidOperator.js";
import { _PolyhedralBoundedSolidSetOperator } from "../_PolyhedralBoundedSolidSetOperator.js";

/**
This class is used to store vertex / halfedge neigborhood information for the
vertex/vertex classifier as proposed on section [MANT1988].15.5. and program
[MANT1988].15.6.
*/
export class _PolyhedralBoundedSolidSetOperatorSectorClassificationOnVertex extends _PolyhedralBoundedSolidOperator {
    public he: _PolyhedralBoundedSolidHalfEdge | null = null;
    public ref1: Vector3Dd | null = null;
    public ref2: Vector3Dd | null = null;
    public ref12: Vector3Dd | null = null;
    public referenceLine: Vector3Dd | null;
    public referenceU: Vector3Dd | null;
    public referenceV: Vector3Dd | null;
    public wide = false;

    public constructor() {
        super();
        this.referenceLine = null;
        this.referenceU = null;
        this.referenceV = null;
    }

    public getAngle(): number {
        if (this.referenceLine === null || this.referenceU === null || this.referenceV === null) {
            return -1000;
        }

        let a = this.ref1!;

        if (_PolyhedralBoundedSolidSetOperator.colinearVectorsWithDirection(this.ref1!, this.referenceLine)) {
            a = this.ref2!;
        }

        let u: Vector3Dd;
        let v: Vector3Dd;

        u = new Vector3Dd(this.referenceU);
        u = u.normalized();
        v = new Vector3Dd(this.referenceV);
        v = v.normalized();
        a = a.normalized();

        const x = a.dotProduct(u);
        const y = a.dotProduct(v);

        let an = Math.acos(x);
        if (y < 0) an *= -1;

        return an;
    }

    public override toString(): string {
        let msg: string;

        msg =
            "R1: " +
            this.ref1 +
            " R2: " +
            this.ref2 +
            " HE " +
            this.he!.startingVertex.id +
            "/" +
            this.he!.next()!.startingVertex.id +
            (this.wide ? "(W)" : "(nw)");

        return msg;
    }

    public compareTo(other: _PolyhedralBoundedSolidSetOperatorSectorClassificationOnVertex): number {
        const a = this.getAngle();
        const b = other.getAngle();

        if (a > b) return 1;
        if (a < b) return -1;
        return 0;
    }
}
